oscduplicator/duplicator.py

```diff
@@ -0,0 +1,248 @@
+import json
+import queue
+import socket
+from dataclasses import asdict
+from threading import Thread
+
+from pythonosc import dispatcher, osc_server, udp_client
+from oscduplicator.setting import TransmitPortSetting
+
+
+class OSCDuplicator:
+    """
+    OSC信号を受信するためのサーバーと、再送信のための設定を保持する
+
+    Attributes
+    ---------
+    server: osc_server.ThreadingOSCUDPServer
+        OSC信号を受信するためのサーバー
+    receive_port: int
+        OSC信号を受信するためのポート、デフォルト9001
+    transmit_port_settings: list[TransmitPortSetting]
+        転送先ポートの設定を保持するリスト
+    clients: list[udp_client.SimpleUDPClient]
+        OSC信号を再送信するための、clientのリスト
+    q: queue.Queue
+        OSC信号の受信順・送信順を保証するためのキュー
+    """
+
+    def __init__(self) -> None:
+        self.__receive_port: int | None = None
+        self.__transmit_port_settings: list[TransmitPortSetting] = []
+        self.__server: osc_server.ThreadingOSCUDPServer | None = None
+        self.clients: list[udp_client.SimpleUDPClient] = []
+        self.__q = queue.Queue()
+
+    @property
+    def receive_port(self):
+        return self.__receive_port
+
+    @receive_port.setter
+    def receive_port(self, value: int):
+        """
+        __receiver_portのsetter
+        値のチェックをする
+
+        Parameters
+        ---------
+        value: int
+            新しい受信ポート番号
+        """
+        # check value
+        if not isinstance(value, int):
+            raise TypeError(
+                "Expected instance of type 'int' for attribute 'port'"
+            )
+        if not (0 <= value <= 65535):
+            raise ValueError(
+                "Invalid 'port' value. Expected a number between 0 and 65535"
+            )
+        self.__receive_port = value
+
+    def start_server(self) -> None:
+        """
+        OSCサーバーを初期化し、起動する
+        """
+        port = self.receive_port if self.receive_port is not None else 9001
+        dpt = dispatcher.Dispatcher()
+
+        # dpt.map("/*", print)
+        dpt.map("/*", self.__queue_osc)
+
+        self.__server = osc_server.ThreadingOSCUDPServer(
+            ("0.0.0.0", port), dpt
+        )
+        self.__server.serve_forever()
+
+    def stop_server(self) -> None:
+        """
+        OSCサーバーを終了する
+        """
+        self.__server.shutdown()
+
+    def __queue_osc(self, address: str, *args: list) -> None:
+        """
+        受信したOSC信号をdictとして、キューに追加する
+
+        Parameters
+        ---------
+        address: str
+            受信・再送信アドレス
+        args: list
+            osc信号
+        """
+        d = {"address": address, "args": args}
+        self.__q.put(d)
+
+    def transmit_msg(self, clients: list[udp_client.SimpleUDPClient]) -> None:
+        """
+        osc信号を転送する
+
+        Parameters
+        ---------
+        transmit_clients: list[udp_client.SimpleUDPClient]
+            OSC信号を再送信するための、clientのリスト
+        q: queue.Queue
+            キュー
+            {"address": address, "args": args}
+        """
+
+        if len(clients) <= 0:
```

Comment:
```suggestion
        if not clients:
```

とも書けます （[空は条件としてはFalse扱い](https://docs.python.org/ja/3/reference/expressions.html#boolean-operations)）